### PR TITLE
Add drop and shutdown for TcpStream

### PR DESCRIPTION
**Pull request by mcches**

### CODE REVIEW

File: src/net/stream.rs [hunk 1]
```diff
@@ -87,31 +98,72 @@ impl TcpStream {
...
         match World::current(|world| world.recv_on(pair)) {
-            Some(seg) => match seg {
-                Segment::Data(bytes) => {
-                    buf.put_slice(bytes.as_ref());
-                    return Some(Ok(()));
-                }
-            },
+            Some(seg) => {
+                let bytes = match seg {
+                    Segment::Data(bytes) => bytes,
```
Comment (LucioFranco) [inline src/net/stream.rs:114]: What happens if a data frame of zero bytes will that trigger some other eof behavior?

**Review by LucioFranco — APPROVED** [review]

File: src/net/stream.rs (hunk 1 above)
Comment (mcches) [reply src/net/stream.rs:114]: Good catch, that is broken. We should no-op the write side to make this impossible.

Comment (mcches) [reply src/net/stream.rs:114]: See [d8a156b](https://github.com/tokio-rs/turmoil/pull/42/commits/d8a156b34d26c7e91ce5b9459a51afc8a3b42354) for fix and test.

**Review by LucioFranco — APPROVED** [review]

**Comment by mcches** [comment]
Rolling this into a bigger refactor.

**Pull request closed without merging**